### PR TITLE
PIL-1214 Register ENS popup

### PR DESCRIPTION
**Pull request by divyaBacancy**

### CODE REVIEW

File: src/screens/RegisterENS/RegisterENS.js [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+import HeaderBlock from 'components/HeaderBlock';
+import Icon from 'components/core/Icon';
+import SwipeButton from 'components/SwipeButton/SwipeButton';
+import Button from 'components/core/Button';
+import FeeLabel from 'components/display/FeeLabel';
+
+// Types
+import type { OnboardingUser } from 'models/User';
+
+
```
Comment (IAmKio) [inline src/screens/RegisterENS/RegisterENS.js:64]:
```suggestion
```

File: src/screens/RegisterENS/RegisterENS.js [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+    //   // console.log('statusCallback', statusCallback)
+    //   navigation.dismiss();
+    //   navigation.navigate(SEND_TOKEN_TRANSACTION, {
+    //     ...transactionStatus,
+    //     noRetry: true,
+    //     goBackDismiss: true,
+    //   });
+    // };
+    // dispatch(sendENSTransaction(statusCallback));
+  };
```
Comment (IAmKio) [inline src/screens/RegisterENS/RegisterENS.js:110]: Can we remove this altogether?

File: src/screens/RegisterENS/RegisterENS.js [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+    // user updated, reset
+    if (isCheckingUsername) setIsCheckingUsername(false);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [user]);
+
+  useEffect(() => {
+    if (!usernameValidationErrorMessage) onValidUsername();
+    return onValidUsername.cancel;
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [onValidUsername, usernameValue]);
```
Comment (IAmKio) [inline src/screens/RegisterENS/RegisterENS.js:147]: Do we need all the separate `useEffect`s? Just wondering if any can be merged together, unless is there a reason for the separation?

File: src/services/etherspot.js [hunk 4]
```diff
@@ -216,6 +217,42 @@ export class EtherspotService {
...
+
+    if (!batch) {
+      reportErrorLog('setTransactionsBatchAndEstimate -> estimateTransactionsBatch returned null', {
+        batch,
+        chain,
+      });
+      return null;
+    }
+
+    // const feeinfo = buildTransactionFeeInfo(batch);
```
Comment (IAmKio) [inline src/services/etherspot.js:252]:
```suggestion
```

File: src/services/etherspot.js [hunk 5]
```diff
@@ -488,6 +525,23 @@ export class EtherspotService {
...
+  ): Promise<?TransactionResult> {
+    const sdk = this.getSdkForChain(chain);
+    if (!sdk) {
+      reportErrorLog('setTransactionsBatchAndSend failed: no SDK for chain set', { chain });
+      throw new Error(t('error.unableToSendTransaction'));
+    }
+
+    // submit current batch
+    const { hash: batchHash } = await sdk.submitGatewayBatch();
+    // console.log('batchHash', batchHash)
```
Comment (IAmKio) [inline src/services/etherspot.js:539]:
```suggestion
```

File: src/services/etherspot.js [hunk 6]
```diff
@@ -488,6 +525,23 @@ export class EtherspotService {
...
+
+    return { batchHash };
+  }
+
```
Comment (IAmKio) [inline src/services/etherspot.js:543]:
```suggestion
```

File: src/services/etherspot.js [hunk 7]
```diff
@@ -216,6 +217,42 @@ export class EtherspotService {
       });
   }
 
+  async estimateBatch(chain: Chain): Promise<?TransactionFeeInfo> {
```
Comment (IAmKio) [inline src/services/etherspot.js:220]: Am i right in thinking that the `estimateBatch` function here is specifically for ENS batch estimation? Could we rename this function to make it ENS specific? The name `estimateBatch` is a bit too ambiguous.

File: src/services/etherspot.js [hunk 8]
```diff
@@ -216,6 +217,42 @@ export class EtherspotService {
...
+    const sdk = this.getSdkForChain(chain);
+    if (!sdk) {
+      reportErrorLog('setTransactionsBatch failed: no SDK for chain set');
+      throw new Error(t('error.unableToSetTransaction'));
+    }
+    const { account: etherspotAccount } = sdk.state;
+    if (isProdEnv() && chain === CHAIN.ETHEREUM && !etherspotAccount?.ensNode) {
+      const ensNode = await this.getEnsNode(etherspotAccount.address);
```
Comment (IAmKio) [inline src/services/etherspot.js:228]: Could we add a `.catch` method here?

File: src/services/etherspot.js [hunk 9]
```diff
@@ -216,6 +217,42 @@ export class EtherspotService {
...
+      if (ensNode && ensNode.state === ENSNodeStates.Reserved) {
+        await sdk.batchClaimENSNode({ nameOrHashOrAddress: ensNode.name });
```
Comment (IAmKio) [inline src/services/etherspot.js:230]: Could we add a `.catch` method here too?

**Review by IAmKio — CHANGES_REQUESTED** [review]
A few changes, but we're nearly there! 👍

File: src/screens/RegisterENS/RegisterENS.js (hunk 3 above)
Comment (divyaBacancy) [reply src/screens/RegisterENS/RegisterENS.js:147]: @IAmKio The dependency array is changed in every useEffect.

**Review by Yash-Chaubal — APPROVED** [review]
LGTM!

**Comment by IAmKio** [comment]
@divyaBacancy I'm happy with all the PR review changes - but can i just confirm that the validation UI changes are as they are in the design? 

020 OF ENS Username accepted
021 Onboarding Username is taken

I'm aware that we changed the swipe button to appear sooner, but if i recall rightly from the screenshots / videos, the name entry / error state needs to be matched up the design.

**Comment by divyaBacancy** [comment]
ENSnamewithFeeandSwipe

**Comment by divyaBacancy** [comment]
https://user-images.githubusercontent.com/66996635/150105640-fd104a5c-c7ef-4191-b97b-584c01483096.mov

File: src/services/etherspot.js [hunk 10]
```diff
@@ -244,26 +245,31 @@ export class EtherspotService {
...
+        try {
+          // parsing etherspot estimate error based on return scheme
+          const errorMessageJson = JSON.parse(error.message.trim());
+          [etherspotErrorMessage] = Object.values(errorMessageJson[0].constraints);
+        } catch (e) {
+          // unable to parse json
+        }
+        const errorMessage = etherspotErrorMessage || error?.message || t('error.unableToEstimateTransaction');
+        reportErrorLog('estimateENSTransactionFee -> estimateGatewayBatch failed', { errorMessage, chain });
+        throw new Error(errorMessage);
```
Comment (IAmKio) [inline src/services/etherspot.js:262]: We'll revisit this later as it's a bit messy

**Review by IAmKio — APPROVED** [review]
Great work!

**Pull request merged**